src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,535 @@ impl Display for SnapKey {
+    impl Write for SnapFile {
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            if self.tmp_file.is_none() {
+                return Ok(0);
+            }
+            let written = try!(self.tmp_file.as_mut().unwrap().0.write(buf));
+            self.digest.write(&buf[..written]);
+            Ok(written)
+        }
+
+        fn flush(&mut self) -> io::Result<()> {
+            if self.tmp_file.is_none() {
+                return Ok(());
             }
+            self.tmp_file.as_mut().unwrap().0.flush()
         }
     }
-}
 
-/// A reader that calculate checksum and verify it without read
-/// it from the beginning again.
-pub struct SnapValidationReader {
-    reader: File,
-    digest: Digest,
-    left: usize,
-    res: Option<u32>,
-}
+    impl SendSnapshotFileCreator for SnapFile {
+        fn create(&mut self, snap: &DbSnapshot, region: &Region) -> raft::Result<()> {
+            if self.exists() {
+                match self.reader().and_then(|mut r| r.validate()) {
+                    Ok(()) => return Ok(()),
+                    Err(e) => {
+                        error!("[region {}] file {} is invalid, will regenerate: {:?}",
+                               region.get_id(),
+                               self.path(),
+                               e);
+                        try!(self.try_delete());
+                        try!(self.init());
+                    }
+                }
+            }
 
-impl SnapValidationReader {
-    /// Open the snap file located at specified path.
-    fn open<P: AsRef<Path>>(path: P) -> io::Result<SnapValidationReader> {
-        let reader = try!(File::open(path.as_ref()));
-        let digest = Digest::new(crc32::IEEE);
-        let len = try!(reader.metadata()).len();
-        if len < CRC32_BYTES_COUNT as u64 {
-            return Err(io::Error::new(ErrorKind::InvalidInput,
-                                      format!("file length {} < {}", len, CRC32_BYTES_COUNT)));
-        }
-        let left = len as usize - CRC32_BYTES_COUNT;
-        Ok(SnapValidationReader {
-            reader: reader,
-            digest: digest,
-            left: left,
-            res: None,
-        })
-    }
-
-    /// Validate the file
-    ///
-    /// If the reader will be consumed after calling this method, no further data can be
-    /// read from this reader again.
-    pub fn validate(&mut self) -> io::Result<()> {
-        if self.res.is_none() {
-            if self.left > 0 {
-                let cap = cmp::min(self.left, DEFAULT_READ_BUFFER_SIZE);
-                let mut buf = vec![0; cap];
-                while self.left > 0 {
-                    try!(self.read(&mut buf));
+            let t = Instant::now();
+            let mut snap_size = 0;
+            let mut snap_key_cnt = 0;
+            let (begin_key, end_key) = (enc_start_key(region), enc_end_key(region));
+            for cf in snap.cf_names() {
```

Comment:
It's not. `CF_RAFT` doesn't contain the range `[begin_key, end_key)`. I prefer exclude `CF_RAFT` explicitly too. I pointed it out long time ago, but @disksing said that it's more simple to just iterate all the cfs.